feat: add support for accessor arrays and refactor `stats/base/variancepn`

AryanBhirud opened the pull request:
Resolves https://github.com/stdlib-js/stdlib/issues/5689 .
## Description

> What is the purpose of this pull request?

This pull request:

 - Adds accessor arrays support.
 - Refactors stats/base/variancepn.

## Related Issues

> Does this pull request have any related issues?

This pull request:

 - resolves [[RFC]: refactor and add protocol support to `stats/base/variancepn` #5689](https://github.com/stdlib-js/stdlib/issues/5689)

## Questions

> Any questions for reviewers of this pull request?

No.
## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Commits referred:
 - stats/base/min: https://github.com/stdlib-js/stdlib/commit/fbb31fbaafbbbf233740a41a88a57f4db1cb427b ([package](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/min))
 - stats/base/cumax: https://github.com/stdlib-js/stdlib/commit/11f1341134c5d6c7c0ecd82c99dff766e08e731a ([package](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/cumax))
 - stats/base/cumin: https://github.com/stdlib-js/stdlib/commit/528efd8000eb7b910f64e63796a8a7c2fdfa03cd ([package](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/cumin))
 - stats/base/mediansorted: https://github.com/stdlib-js/stdlib/commit/065f86535e3f1915160f91820f1fa0af79633b72 ([package](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/mediansorted))
 - stats/base/mskmax: https://github.com/stdlib-js/stdlib/commit/417e65361234ec46382f58cc303a544f5bbcbf9e ([package](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/mskmax))

I have gone line-by-line in [@stdlib/stats/base/variancepn](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/variancepn) and compared with the changes made in the above commits in order to refactor and add accessor protocol support.

Modifed the lib/ndarray.js file to delegate to the accessors implementation when one (or more) of the input arrays is an accessor array.

Modified the tests to include explicit tests for accessor arrays.
## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

@stdlib-js/reviewers

AryanBhirud committed:
feat: add accessor protocol support to

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

AryanBhirud commented:
@kgryte could you please review my PR

kgryte committed:
Discard changes to CONTRIBUTORS

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/repl.txt:
```diff
@@ -45,21 +45,20 @@
+    // Using `N` and stride parameters:
     > x = [ -2.0, 1.0, 1.0, -5.0, 2.0, -1.0 ];
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
     > var stride = 2;
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/repl.txt:
```diff
@@ -45,21 +45,20 @@
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
     > var stride = 2;
-    > {{alias}}( N, 1, x, stride )
+    > {{alias}}( 3, 1, x, stride )
```
```suggestion
    > {{alias}}( 3, 1, x, 2 )
```

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/repl.txt:
```diff
@@ -45,21 +45,20 @@
     > var x0 = new {{alias:@stdlib/array/float64}}( [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ] );
     > var x1 = new {{alias:@stdlib/array/float64}}( x0.buffer, x0.BYTES_PER_ELEMENT*1 );
-    > N = {{alias:@stdlib/math/base/special/floor}}( x0.length / 2 );
     > stride = 2;
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/repl.txt:
```diff
@@ -45,21 +45,20 @@
-    > N = {{alias:@stdlib/math/base/special/floor}}( x0.length / 2 );
     > stride = 2;
-    > {{alias}}( N, 1, x1, stride )
+    > {{alias}}( 3, 1, x1, stride )
```
```suggestion
    > {{alias}}( 3, 1, x1, 2 )
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/types/index.d.ts:
```diff
@@ -20,14 +20,20 @@
 interface Routine {
 	/**
-	* Computes the variance of a strided array using a two-pass algorithm.
+	* Computes the variance of a strided array using a two-pass algorithm and alternative indexing semantics.
```
```suggestion
	* Computes the variance of a strided array using a two-pass algorithm.
```

kgryte committed:
docs: fix description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update parameter names

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/types/index.d.ts:
```diff
@@ -32,25 +38,25 @@ interface Routine {
 	* var x = [ 1.0, -2.0, 2.0 ];
 	*
-	* var v = variancepn( x.length, 1, x, 1 );
+	* var v = variancepn( x.length, 1, x, 1, 0 );
```
```suggestion
	* var v = variancepn( x.length, 1, x, 1 );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/types/index.d.ts:
```diff
@@ -32,25 +38,25 @@ interface Routine {
 	* // returns ~4.3333
 	*/
-	( N: number, correction: number, x: NumericArray, stride: number ): number;
+	( N: number, correction: number, x: NumericArray | InputArray, strideX: number ): number;
```
```suggestion
	( N: number, correction: number, x: InputArray, strideX: number ): number;
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
Discard changes to lib/node_modules/@stdlib/stats/base/variancepn/test/test.js

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feat/variancepn-accessor-support

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feat/variancepn-accessor-support

gururaj1512 committed:
feat: refactor variancepn implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: correct copyright year in variancepn example

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: update copyright year in benchmark files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: variancepn test files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: correct variancepn README

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 reviewed (dismissed):
LGTM after cleanup. Onto @kgryte

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/docs/repl.txt:
```diff
@@ -108,8 +105,7 @@
 
     // Using offset parameter:
     > var x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ];
```
```suggestion
    > x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ];
```

kgryte committed:
docs: avoid duplicate declaration

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/lib/accessors.js:
```diff
@@ -59,29 +66,31 @@ function variancepn( N, correction, x, stride ) {
+	// Cache a reference to the element accessor:
+	get = x.accessors[ 0 ];
+
+	if ( N === 1 || strideX === 0 ) {
 		return 0.0;
 	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/variancepn/lib/ndarray.js:
```diff
@@ -66,21 +66,26 @@ function variancepn( N, correction, x, stride, offset ) {
+	if ( N === 1 || strideX === 0 ) {
 		return 0.0;
 	}
+
```
```suggestion
	if ( N === 1 || strideX === 0 ) {
		return 0.0;
	}
	o = arraylike2object( x );
	if ( o.accessorProtocol ) {
		return accessors( N, correction, o, strideX, offsetX );
	}
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
